sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -389,6 +389,23 @@ case class DataSource(
   }
 
   /**
+   * Return all paths represented by the wildcard string.
+   */
+  private def getGlobbedPaths(qualified: Path): Seq[Path] = {
```

Comment:
at least we should follow `InMemoryFileIndex.bulkListLeafFiles` and `Picks the listing strategy adaptively depending on the number of paths to list`